library/core/src/escape.rs

```diff
@@ -149,77 +151,197 @@ const fn escape_unicode<const N: usize>(c: char) -> ([ascii::Char; N], Range<u8>
     (output, (start as u8)..(N as u8))
 }
 
+#[derive(Clone, Copy)]
+union MaybeEscapedCharacter<const N: usize> {
+    pub escaped: [ascii::Char; N],
+    pub unescaped: char,
+}
+
+/// Marker type to indicate that the character is always escaped,
+/// use to optimized the iterator implementation.
+#[derive(Clone, Copy)]
+#[non_exhaustive]
+pub(crate) struct AlwaysEscaped;
+
+/// Marker type to indicate that the character may be escaped,
+/// use to optimized the iterator implementation.
+#[derive(Clone, Copy)]
+#[non_exhaustive]
+pub(crate) struct MaybeEscaped;
+
 /// An iterator over an fixed-size array.
 ///
-/// This is essentially equivalent to array’s IntoIter except that indexes are
-/// limited to u8 to reduce size of the structure.
-#[derive(Clone, Debug)]
-pub(crate) struct EscapeIterInner<const N: usize> {
+/// This is essentially equivalent to array’s `IntoIter` except that
+/// indices are stored as `u8` to reduce size of the structure.
+#[derive(Clone)]
+pub(crate) struct EscapeIterInner<const N: usize, ESCAPING> {
     // The element type ensures this is always ASCII, and thus also valid UTF-8.
-    data: [ascii::Char; N],
+    // Invariant: `N < 128`. For non-printable characters, contains the escaped
+    // representation as ASCII characters. For printable characters, contains the
+    // character itself.
+    data: MaybeEscapedCharacter<N>,
 
-    // Invariant: `alive.start <= alive.end <= N`
+    // Invariant: For non-printable characters, `alive.start <= alive.end <= N`,
+    // for printable characters, `alive.end >= alive.start >= 128`.
     alive: Range<u8>,
+
+    escaping: PhantomData<ESCAPING>,
 }
 
-impl<const N: usize> EscapeIterInner<N> {
+impl<const N: usize, ESCAPING> EscapeIterInner<N, ESCAPING> {
+    #[inline]
+    const fn new(data: MaybeEscapedCharacter<N>, alive: Range<u8>) -> Self {
+        const { assert!(N < 128) };
+        Self { data, alive, escaping: PhantomData }
+    }
```

Comment:
This does not seem to uphold the invariant of this type. What prevents a caller from calling this with an alive range of 0..255 for example?

Reply:
Nothing. Made it `unsafe` now.